Draft - Http cache / rule rework for next major

OliverSkroblin opened the pull request:
This merge requests reflects the current work in progress state of the cache/rule system rework which also has a lot of dependencies and many different tasks.

The topics are already discussed in many discussions, adrs or other merge requests:
- https://github.com/shopware/shopware/discussions/3489
- https://github.com/shopware/shopware/discussions/3165
- https://github.com/shopware/shopware/discussions/3406
- https://github.com/shopware/shopware/discussions/3391

I will try to split all commits into separated topic and keep the task list up to date.

I publish it here on github to have it public as fast as possible and clear communicated for all of you. And start discussing different solutions.

Goal of the changes is:
- Reduce global state complexity 
- Increadible performance boost and cache hit improvement (now: ~5%, goal: 100%)
- Less cache purges
- Reduce indexing ressources (message queue overload right now)
- Better extendability 
- Clear and documented APIs for important business classes

Here the task list:
# General
- [ ] Rector rules for easy update
- [ ] Upgrade guide
- [ ] Add profiler element which tags would be set for the current page

# Context rules
- [x] Rules are no longer available in context object
- [x] Rewrite ActiveRulesDataCollectorSubscriber and show rules of the cart (if cart was loaded) > maybe via event + listener?
- [x] Remove usage of area rule ids 

# Cart rules
- [x] Rules in cart will be validated after each processor
- [x] Attach matching rules to cart
- [x] Adapt classes in cart domain which rely on context.rules and change it to cart.rules
- [x] Remove cart rule loader
 
# product (pricing) api
- [ ] New php class to load the product prices
- [ ] Different scopes listing/pdp/cart
- [ ] New pricing concept (customer group, sales channel, delivery_country?)
- [ ] Rework of cheapest price updater
- [ ] rework the cheapest price accessor builder in sql
- [ ] Rework the cheapest price in elasticsearch
- [ ] Price templates 
   - [ ] storefront 
   - [ ] feeds
   - [ ] exports
- [x] Implement new event to invalidate product cache
- [x] provide new http api endpoint to import product prices
- [x] instant purge on http cache for out-of-stock/price-changed products
- [x] New pricing concept
   - [x] customer group
   - [x] customer group + channel
   - [x] customer group + channel + country
   - [x] % discount
   
# Cart widget
- [x] Store cart widget in local storage for X minutes and refresh on cart actions (add, delete, change)

# Wishlist widget
- [x] Just fetch the count
- [x] Cache result in local storage

# http cache 
- [x] Remove rules from cache key
- [x] Dispatch event to calculate http cache key
- [x] Add cookie list to allow key extension
- [x] Catch event in CacheStore and Reverse proxy
- [x] Remove cache decorator
- [x] Rework cache invalidator, to optimize tags and hit invalidation count
- [x] New route to invalidate "delay" cache
- [x] Refactor admin shortcuts for cache delete, maybe also remove "delete all caches" 
- [x] Remove cart access in cache response subscriber, otherwise we always access the cart again to set the states (cart-filled, etc)
- [x] Hide `AddCacheTagEvent` dispatching behind feature flag
- [x] Global cms-page tagging handling for cache invalidation


# Store api
- [x] Use new `AddCacheTagEvent` event to add tags to http cache
- [ ] Store api full page cache
- [x] Remove cached routes
- [ ] I would like to remove abstract routes and integrate new event-decoration pattern 
- [ ] Remove route decorations (Resolved...), there is no more cache layer between so we don't need to resolve the request before calling the "real" route

# delay cache 
- [x] Activate delay cache by default
- [x] Delay everything
- [ ] Shortcut in admin for clear + dropdown(?) button in cms/category/product module to save+invalidate-delayed
 
# Esi integration
- [x] Header and footer via esi 
- [x] header pagelet no more in page
- [x] Replace activeLanguage 
- [x] Replace activeCurrency
- [x] Replace minSearchLength
- [ ] Cms category sidebar element own resolver
- [x] Replace active navigation id
- [x] Add http cache support esi requests
- [x] Breadcrumb via id
- [x] Remove header from page object
- [x] Remove footer from page object
- [ ] Reduce data shipping and payment method for footer
- [x] Remove caching of shipping method / payment method route tag
- [x] Remove header.serviceMenu (and just use footer.serviceMenu)

# Navigation route
- [x] Better cache invalidation
- [x] Break navigation route to just provide main, service, footer function
- [x] Reduce data which is returned

# Flow improvement
- [ ] Flow can no more rely on context.ruleIds
- [ ] Should also not rely on order object loaded
- [ ] Rules in flows should be allowed to query data by their own
- [ ] Remove big data objects from flow data and migrate rules to new rule system
- [ ] Dont restore context / order in flows to evaluate rules

# Theme caching
- [x] Rework theme caching to have one tag for all theme related stuff because of shared on each page
- [x] Includes theme config, theme assignment, snippets, etc

---

# Nice to have

# Twig rendering performance
- [ ] Improve filter panel rendering
- [ ] Improve product list rendering
- [ ] Improve navigation rendering

## Document storage
- [ ] Require new document storage as composer dependency
- [ ] Use document storage for product listing > At least minimal data stack for filter
- [ ] Use document storage for navigation > At least minimal data stack for navigation 

## Decorator extension
- [ ] Introduce new decoration pattern via event

## Documentation system
- [ ] Introduce new documentation system
- [ ] Added documentation for new store-api route decorations


# Pricing concept
![price-concept](https://github.com/shopware/shopware/assets/1035358/f130f19a-48ed-4959-8090-7f62a79f1592)

# Flow Concept
![image](https://github.com/shopware/shopware/assets/1035358/0ad2b9d9-19a7-4bd1-b5ef-2ea8e05c27ca)

ssltg commented on src/Storefront/Framework/Twig/TemplateDataExtension.php:
```diff
@@ -75,4 +80,36 @@ private function getControllerInfo(Request $request): ControllerInfo
+        return $min ? (int) $min : 3;
+    }
+
+    private function getLanguage(SalesChannelContext $context): array
```
The language informations belong to the SaleschannelContext, as the currencies do.
Also minSearchLength seam to be a bit random to be selected as "always available". Either we have an object that holds bearly changed standard configurations or we get them when we need them, but select single configs one after one to be stored in the twigContext does make things more complex and unreliable, IMHO.

OliverSkroblin replied:
IMO that's not correct. The sales channel context was designed for data which is required in the PHP stack globally. The TwigContext, or DataExtensions, are designed for global data which are required inside Twig. 
All data i selected here is just neccessary for twig and not in the php stack. 

And because the data is spreaded in base templates and headers, i just moved it here.

mitelg commented on src/Storefront/Controller/EsiController.php:
```diff
@@ -0,0 +1,41 @@
+ */
+#[Route(defaults: ['_routeScope' => ['storefront']])]
+#[Package('storefront')]
+class EsiController extends StorefrontController
```
I would suggest, to not introduce one global ESI controller. instead having a controller for each header and footer. and then additional controller for the ESI things to come in the future. 

or is there any advantage having all route in one (going to be big) controller class?

ssltg replied:
I would suggest to not introduce special endpoints for esi at all. render_esi should just call the endpoint that calls the pageletLoader. If this is handled with esi or not is an configuration detail.

OliverSkroblin replied:
But right now we have no controller endpoint which "just calls" the pagelet loader.

ssltg replied:
Then implement one

OliverSkroblin replied:
??? sorry i don't get it. Don't see the need to implement another controller right now? I added this controller which just contains 2 lines of real code and they are always calling just the pagelet loaders and render the pagelet.

Don't see any reason right now to adjust the code here.

ssltg replied:
My point is, that there is no need to do a dedicated ESI controller if we can have the same (same number of lines and code) just with a name that don't force a special kind of layer. A HeaderController and FooterController with the same methods and not "esi" in its path does not add more code or complexity but more freedom of the exact kind of implementation.

OliverSkroblin replied:
But keep in mind, the route has to be marked as `_esi`

```
#[Route(path: '/esi/header', defaults: ['_esi' => true], .....
```

Which also means, this route should only be called via ESI because the sub request handling is another then a normal request

ssltg replied:
But it could also called by something "like esi" for instance if we implement some kind of defered loading, or would there be a problem then?

OliverSkroblin replied:
don't think so. I mean we had a check for this _esi (earlier) but i can not find it anymore.

mitelg commented on src/Storefront/Pagelet/Header/HeaderPageletLoader.php:
```diff
@@ -108,6 +108,8 @@ private function getLanguages(SalesChannelContext $context, Request $request): L
 
         $criteria->addSorting(new FieldSorting('name', FieldSorting::ASCENDING));
+
+        // todo@skroblin why is this loaded here???
```
because the search is in the header? :shrug: 

https://github.com/shopware/shopware/blob/trunk/src/Storefront/Resources/views/storefront/layout/header/search.html.twig#L2

OliverSkroblin replied:
yey, but i moved it into the template data extension.

mitelg replied:
then, it is probably not needed anymore here :blush:

OliverSkroblin committed:
NEXT-31669 - Added esi decoration for native http cache support

OliverSkroblin committed:
NEXT-31669 - Introduce new invalidate product cache event for force clear

OliverSkroblin committed:
NEXT-31669 - Implement new symfony toolbar section to see http cache tags

OliverSkroblin committed:
NEXT-31669 - Fixed nested and force invalidation of theme and translator cache

OliverSkroblin committed:
NEXT-31669 - Update service definitions

OliverSkroblin committed:
NEXT-31669 - Get rid of global context / page variables and use template data extension for this

OliverSkroblin committed:
NEXT-31669 - Remove cached*route layer and use new http cache tag event for tagging

OliverSkroblin committed:
NEXT-31669 - Remove none-delayed cache system and optimize tagging and invalidation

OliverSkroblin commented:
Branch was reduced to just ESI + Cache invalidation rework.

Rule + product rework moved to: https://github.com/shopware/shopware/pull/3729

OliverSkroblin committed:
NEXT-31669 - New command to watch delayed cache storage

OliverSkroblin committed:
NEXT-31669 - New sw_breadcrumb_full_by_id function

OliverSkroblin committed:
NEXT-31669 - Use render_esi to load and render navigation and footer. Also implement new api endpoints for header/footer/service menu to remove complexity and improve cache hit rate

NEXT-31669 - Added changelog and upgrade information. Fix tests and code style.

NEXT-31669 - OffCanvas menu initial content via ajax
